Python/ceval.c

```diff
@@ -4107,14 +4107,13 @@ _PyEval_EvalCode(PyThreadState *tstate,
 {
     assert(is_tstate_valid(tstate));
 
-    PyCodeObject* co = (PyCodeObject*)_co;
-    PyFrameObject *f;
+    PyCodeObject *co = (PyCodeObject*)_co;
+    PyObject *code_name = qualname ? qualname : co->co_name;
```

Comment:
Rather than putting this "fix" in `_PyEval_EvalCode()`, what about the idea of documenting that the `qualname` (or `code_name`) argument can't be NULL (and asserting that), and updating the callers instead? There are only a couple callers, and it will put the messiness closer to the source of the issue.

Reply:
qualname **is** NULL in `PyEval_EvalCodeEx`. I don't think that it would be possible to pass `co->co_name` as `qualname` in such case.

Reply:
`qualname` parameter of `PyEval_EvalCode()` is passed to `PyCoro_New()` for example. Extract of its code:

```
    if (name != NULL)
        gen->gi_name = name;
    else
        gen->gi_name = ((PyCodeObject *)gen->gi_code)->co_name;
    Py_INCREF(gen->gi_name);
    if (qualname != NULL)
        gen->gi_qualname = qualname;
    else
        gen->gi_qualname = gen->gi_name;
    Py_INCREF(gen->gi_qualname);
```

Note: In `PyEval_EvalCodeEx()`, `name` is also NULL.

Or maybe we can set name and qualname to a more useful value than NULL in `PyEval_EvalCodeEx()`, I don't know.

Reply:
Yeah, that's why I was saying "(or `code_name`)" in my comment (i.e. `_PyEval_EvalCode()`'s argument can be changed to `code_name`). I'm not sure the "fix" is needed outside of `PyEval_EvalCodeEx()`, which is why I was suggesting it would maybe be better there.

Reply:
(I don't know the implications for `PyCoro_New()` though.)

Reply:
Even if it's "private" (``_Py``) prefix, _PyEval_EvalCodeWithName() is exposed in the C API and **can** be called with `qualname=NULL`. So I don't think that it's worth it to attempt to leave `_PyEval_EvalCode()` unchanged: I consider that `_PyEval_EvalCode()` is the right place to fix the issue.

Anyway, I wrote PR #20615 which tries to change less code.

Reply:
Thanks for discussing with me. I approved the other PR, but either is fine (not sure which you prefer at this point).